redis/subConn.go

```diff
@@ -78,55 +80,100 @@ func (c *subConn) ReceiveChan() <-chan redis.Message {
 
 func (c *subConn) mainLoop() {
 	var (
-		pingTicker      = time.NewTicker(pingDelay)
+		pingTicker      = time.Tick(pingDelay)
 		pongTimeoutChan <-chan time.Time
 
 		done    = make(chan struct{})
 		msgChan = connReceiveChan(c.currConn, done)
 	)
-	recover := func() {
+	recoverConn := func() {
```

Comment:
We actually cannot, since the trick is that this function alters some of the local state of the closure here, which are all the variables defined above (I think only a few of them are actually immutable).

We also cannot share this on the package level otherwise it would be shared between several instances of the `subConn` struct, which would create problems if any consumer has more than 1 instance of it.

I could create a separate struct which contained this "loop state" so we can have some methods on it and modify only the state respective to our specific loop, but I think this might be overcomplicating it a bit. WDYT?

Reply:
Sorry, my brain skipped the fact that `mainLoop` is a method and not a "pure" function.
I think you may replace `package` by `subConn` in my comments.

The idea of having a connection state `struct` is not that bad either. It would also simplify the parameter list that `mainIteration` takes. Then `recorverConn` would be a top level function that receives such `struct`. It keeps the locality and also groups all stuff related to the connection state.

Mostly a ⛏️  though. Feel free o ignore it.

Reply:
Done! Could you take another look? I didn't do any `push -f`, so you can check only the new commits after you reviewed :)